File: ios/RoutineChart/RoutineChart/Core/Data/Remote/Firebase/FirestoreFamilyInviteSyncService.swift
//
//  FirestoreFamilyInviteSyncService.swift
//  RoutineChart
//
//  Created for Phase 2.3.3: Firestore Sync (Invites)
//

import Foundation
import FirebaseFirestore
import OSLog

/// Service for syncing FamilyInvite data to/from Firestore
/// MIGRATED: Now uses top-level /invites collection for simpler queries
final class FirestoreFamilyInviteSyncService {
    private let db = Firestore.firestore()
    
    /// Top-level invites collection: /invites/{inviteId}
    private func invitesCollection() -> CollectionReference {
        db.collection("invites")
    }
    
    /// Sync invite to Firestore
    func syncToFirestore(_ invite: FamilyInvite) async throws {
        let inviteData: [String: Any] = [
            "id": invite.id,
            "familyId": invite.familyId,
            "token": invite.token,
            "inviteCode": invite.inviteCode,
            "createdBy": invite.createdBy,
            "createdAt": Timestamp(date: invite.createdAt),
            "expiresAt": Timestamp(date: invite.expiresAt),
            "maxUses": invite.maxUses as Any,
            "usedCount": invite.usedCount,
            "isActive": invite.isActive
        ]
        
        try await invitesCollection()
            .document(invite.id)
            .setData(inviteData, merge: true)
        
        AppLogger.database.info("Synced invite to Firestore: \(invite.id)")
    }
    
    /// Increment usedCount for an invite (optimized for security rules)
    /// Uses updateData to only update the usedCount field, which works better with security rules
    func incrementUsedCount(inviteId: String, newUsedCount: Int) async throws {
        let updateData: [String: Any] = [
            "usedCount": newUsedCount
        ]
        
        try await invitesCollection()
            .document(inviteId)
            .updateData(updateData)
        
        AppLogger.database.info("Incremented usedCount for invite \(inviteId) to \(newUsedCount)")
    }
    
    /// Sync invites from Firestore for a family
    func syncFromFirestore(familyId: String) async throws -> [FamilyInvite] {
        let snapshot = try await invitesCollection()
            .whereField("familyId", isEqualTo: familyId)
            .getDocuments(source: .server)  // Force server read to bypass cache
        
        var invites: [FamilyInvite] = []
        for document in snapshot.documents {
            guard let invite = try? parseInvite(from: document.data(), id: document.documentID) else {
                AppLogger.database.error("Failed to parse invite from Firestore: \(document.documentID)")
                continue
            }
            invites.append(invite)
        }
        
        AppLogger.database.info("Synced \(invites.count) invites from Firestore for family: \(familyId)")
        return invites
    }
    
    /// Query Firestore for an invite by invite code.
    /// Uses top-level collection with direct query (much faster than collection group).
    /// Uses .server source to force network request and bypass cache.
    func getByInviteCodeFromFirestore(_ inviteCode: String) async throws -> FamilyInvite? {
        AppLogger.database.info("Querying Firestore (server) for invite code: \(inviteCode)")
        let snapshot = try await invitesCollection()
            .whereField("inviteCode", isEqualTo: inviteCode)
            .limit(to: 1)
            .getDocuments(source: .server)  // Force server read to bypass cache
        
        guard !snapshot.documents.isEmpty else {
            AppLogger.database.info("No invite found with code: \(inviteCode)")
            return nil
        }
        
        let document = snapshot.documents.first!
        let invite = try parseInvite(from: document.data(), id: document.documentID)
        AppLogger.database.info("Found invite \(invite.id) for code: \(inviteCode)")
        return invite
    }
    
    /// Query Firestore for an invite by token.
    /// Uses top-level collection with direct query (much faster than collection group).
    /// Uses .server source to force network request and bypass cache.
    func getByTokenFromFirestore(_ token: String) async throws -> FamilyInvite? {
        AppLogger.database.info("Querying Firestore (server) for invite token: \(token)")
        let snapshot = try await invitesCollection()
            .whereField("token", isEqualTo: token)
            .limit(to: 1)
            .getDocuments(source: .server)  // Force server read to bypass cache
        
        guard !snapshot.documents.isEmpty else {
            AppLogger.database.info("No invite found with token: \(token)")
            return nil
        }
        
        let document = snapshot.documents.first!
        let invite = try parseInvite(from: document.data(), id: document.documentID)
        AppLogger.database.info("Found invite \(invite.id) for token: \(token)")
        return invite
    }
    
    /// Parse Firestore document data into FamilyInvite
    private func parseInvite(from data: [String: Any], id: String) throws -> FamilyInvite {
        guard let familyId = data["familyId"] as? String,
              let token = data["token"] as? String,
              let inviteCode = data["inviteCode"] as? String,
              let createdBy = data["createdBy"] as? String,
              let createdAtTimestamp = data["createdAt"] as? Timestamp,
              let expiresAtTimestamp = data["expiresAt"] as? Timestamp,
              let usedCount = data["usedCount"] as? Int,
              let isActive = data["isActive"] as? Bool else {
            throw SyncError.invalidData("Missing required fields for FamilyInvite")
        }
        
        let createdAt = createdAtTimestamp.dateValue()
        let expiresAt = expiresAtTimestamp.dateValue()
        let maxUses = data["maxUses"] as? Int
        
        return FamilyInvite(
            id: id,
            familyId: familyId,
            token: token,
            inviteCode: inviteCode,
            createdBy: createdBy,
            createdAt: createdAt,
            expiresAt: expiresAt,
            maxUses: maxUses,
            usedCount: usedCount,
            isActive: isActive
        )
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/data/remote/firebase/FirestoreFamilyInviteSyncService.kt
package com.HammersTech.RoutineChart.core.data.remote.firebase

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.HammersTech.RoutineChart.core.domain.models.FamilyInvite
import com.HammersTech.RoutineChart.core.utils.AppLogger
import kotlinx.coroutines.tasks.await
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for syncing FamilyInvite data to/from Firestore
 * MIGRATED: Now uses top-level /invites collection for simpler queries
 */
@Singleton
class FirestoreFamilyInviteSyncService @Inject constructor() {
    
    private val db = FirebaseFirestore.getInstance()
    
    /**
     * Top-level invites collection: /invites/{inviteId}
     */
    private fun invitesCollection() = db.collection("invites")
    
    /**
     * Sync invite to Firestore. Uses set(merge: true) for upsert.
     */
    suspend fun syncToFirestore(invite: FamilyInvite) {
        try {
            val createdAtTimestamp = com.google.firebase.Timestamp(java.util.Date.from(invite.createdAt))
            val expiresAtTimestamp = com.google.firebase.Timestamp(java.util.Date.from(invite.expiresAt))
            
            val data = hashMapOf<String, Any>(
                "id" to invite.id,
                "familyId" to invite.familyId,
                "token" to invite.token,
                "inviteCode" to invite.inviteCode,
                "createdBy" to invite.createdBy,
                "createdAt" to createdAtTimestamp,
                "expiresAt" to expiresAtTimestamp,
                "usedCount" to invite.usedCount,
                "isActive" to invite.isActive
            )
            
            invite.maxUses?.let { data["maxUses"] = it }
            
            invitesCollection()
                .document(invite.id)
                .set(data, com.google.firebase.firestore.SetOptions.merge())
                .await()
            
            AppLogger.Database.info("Synced invite to Firestore: ${invite.id}")
        } catch (e: Exception) {
            AppLogger.Database.error("Failed to sync invite to Firestore", e)
            throw e
        }
    }
    
    /**
     * Fetches all invites for a given family from Firestore.
     * Uses Source.SERVER to force network request and bypass cache.
     */
    suspend fun syncFromFirestore(familyId: String): List<FamilyInvite> {
        return try {
            AppLogger.Database.info("Fetching invites from Firestore (server) for family: $familyId")
            val snapshot = invitesCollection()
                .whereEqualTo("familyId", familyId)
                .get(Source.SERVER)  // Force server read to bypass cache
                .await()
            
            val invites = snapshot.documents.mapNotNull { document ->
                try {
                    parseInvite(document.data ?: emptyMap(), document.id)
                } catch (e: Exception) {
                    AppLogger.Database.error("Error parsing Firestore invite document ${document.id}", e)
                    null
                }
            }
            AppLogger.Database.info("Successfully fetched ${invites.size} invites from Firestore for family: $familyId")
            invites
        } catch (e: Exception) {
            AppLogger.Database.error("Failed to sync invites from Firestore for family: $familyId", e)
            // Log specific error details
            when {
                e.message?.contains("network", ignoreCase = true) == true -> {
                    AppLogger.Database.error("Network error detected. Check internet connection and Firebase configuration.")
                }
                e.message?.contains("permission", ignoreCase = true) == true -> {
                    AppLogger.Database.error("Permission denied. Check Firestore security rules.")
                }
                else -> {
                    AppLogger.Database.error("Unknown error: ${e.message}")
                }
            }
            throw e
        }
    }
    
    /**
     * Query Firestore for an invite by invite code.
     * Uses top-level collection with direct query (much faster than collection group).
     * Uses Source.SERVER to force network request and bypass cache.
     */
    suspend fun getByInviteCodeFromFirestore(inviteCode: String): FamilyInvite? {
        return try {
            AppLogger.Database.info("Querying Firestore (server) for invite code: $inviteCode")
            val snapshot = invitesCollection()
                .whereEqualTo("inviteCode", inviteCode)
                .limit(1)
                .get(Source.SERVER)  // Force server read to bypass cache
                .await()
            
            if (snapshot.isEmpty) {
                AppLogger.Database.info("No invite found with code: $inviteCode")
                return null
            }
            
            val document = snapshot.documents.first()
            val invite = parseInvite(document.data ?: emptyMap(), document.id)
            AppLogger.Database.info("Found invite ${invite.id} for code: $inviteCode")
            invite
        } catch (e: Exception) {
            AppLogger.Database.error("Failed to query invite by code from Firestore: $inviteCode", e)
            throw e
        }
    }
    
    /**
     * Query Firestore for an invite by token.
     * Uses top-level collection with direct query (much faster than collection group).
     * Uses Source.SERVER to force network request and bypass cache.
     */
    suspend fun getByTokenFromFirestore(token: String): FamilyInvite? {
        return try {
            AppLogger.Database.info("Querying Firestore (server) for invite token: $token")
            val snapshot = invitesCollection()
                .whereEqualTo("token", token)
                .limit(1)
                .get(Source.SERVER)  // Force server read to bypass cache
                .await()
            
            if (snapshot.isEmpty) {
                AppLogger.Database.info("No invite found with token: $token")
                return null
            }
            
            val document = snapshot.documents.first()
            val invite = parseInvite(document.data ?: emptyMap(), document.id)
            AppLogger.Database.info("Found invite ${invite.id} for token: $token")
            invite
        } catch (e: Exception) {
            AppLogger.Database.error("Failed to query invite by token from Firestore: $token", e)
            throw e
        }
    }
    
    /**
     * Parse Firestore document data into FamilyInvite
     */
    private fun parseInvite(data: Map<String, Any>, id: String): FamilyInvite {
        val familyId = data["familyId"] as? String
            ?: throw IllegalArgumentException("Missing familyId")
        val token = data["token"] as? String
            ?: throw IllegalArgumentException("Missing token")
        val inviteCode = data["inviteCode"] as? String
            ?: throw IllegalArgumentException("Missing inviteCode")
        val createdBy = data["createdBy"] as? String
            ?: throw IllegalArgumentException("Missing createdBy")
        val createdAtTimestamp = data["createdAt"] as? com.google.firebase.Timestamp
            ?: throw IllegalArgumentException("Missing createdAt")
        val expiresAtTimestamp = data["expiresAt"] as? com.google.firebase.Timestamp
            ?: throw IllegalArgumentException("Missing expiresAt")
        val usedCount = (data["usedCount"] as? Number)?.toInt()
            ?: throw IllegalArgumentException("Missing usedCount")
        val isActive = data["isActive"] as? Boolean
            ?: throw IllegalArgumentException("Missing isActive")
        
        val createdAt = createdAtTimestamp.toDate().toInstant()
        val expiresAt = expiresAtTimestamp.toDate().toInstant()
        val maxUses = (data["maxUses"] as? Number)?.toInt()
        
        return FamilyInvite(
            id = id,
            familyId = familyId,
            token = token,
            inviteCode = inviteCode,
            createdBy = createdBy,
            createdAt = createdAt,
            expiresAt = expiresAt,
            maxUses = maxUses,
            usedCount = usedCount,
            isActive = isActive
        )
    }
}
